Address UTF-8 Detection In Get-Content -Tail

NoMoreFood opened the pull request:
# PR Summary

Fix #11830
Addresses a comparison failure that causes UTF-8 detection to fail which in turn causes Get-Content -Tail to resort to forward lookups given encoding type cannot be detected. Possible this misdetection is due to the incoming encoding object as being of type System.Text.UTF8Encoding where as the comparison uses the object Encoding.UTF8 which is derived from System.Text.UTF8Encoding+UTF8EncodingSealed.

## PR Context

Problem was discovered when investigating performance issues for Get-Content -Tail, in general.  The problem was narrowed to the fact that -Tail will read the entire file when using Get-Content on a UTF-8 file.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed:
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

NoMoreFood committed:
Address UTF-8 Detection In Get-Content -Tail

- Addresses a comparison failure that causes UTF-8 detection to fail which in turn causes Get-Content -Tail to resort to forward lookups given encoding type cannot be detected. Possible this misdetection is due to the incoming encoding object as being of type System.Text.UTF8Encoding where as the comparison uses the object Encoding.UTF8 which is derived from System.Text.UTF8Encoding+UTF8EncodingSealed.
- See https://github.com/PowerShell/PowerShell/issues/11830

iSazonov commented:
@NoMoreFood Please add a test.

vexx32 commented on src/System.Management.Automation/namespaces/FileSystemContentStream.cs:
```diff
@@ -1460,10 +1460,9 @@ private int RefillByteBuff()
+            else if (_currentEncoding is UnicodeEncoding ||
+                _currentEncoding is UTF32Encoding ||
+                _currentEncoding is ASCIIEncoding ||
                 IsSingleByteCharacterSet())
```
Is there a reason to remove the `BigEndianUnicode` handling here?

NoMoreFood replied:
Encoding.Unicode and Encoding.BigEndianUnicode both inherit from UnicodeEncoding so the one test handles both.

NoMoreFood commented:
@iSazonov Since this addresses an internal performance issue, did you just want a textual output posted here with before/after performance tests for a variety of text encodings?

iSazonov commented:
@NoMoreFood I do not look the issue in depths but I feel that if UTF-8 detection wrong the cmdlet can return wrong results. In the case we should add tests.

NoMoreFood commented:
@NoMoreFood Alright, I'll take a look into creating a general, functional test.  I will say though that this was not an issue on Windows PowerShell.  And notably, Windows PowerShell was not failing the `_currentEncoding.Equals(Encoding.UTF8)` comparison that resulted in this problem.

NoMoreFood commented:
@iSazonov Alright, I'll take a look into creating a general, functional test. I will say though that this was not an issue on Windows PowerShell. And notably, Windows PowerShell was not failing the _currentEncoding.Equals(Encoding.UTF8) comparison that resulted in this problem.

vexx32 commented:
I imagine there may have been a change in the .NET Core API at some point that resulted in that failing. It seems on the surface like a relatively innocuous change until you start doing reference equality comparisons against it for things like this. 😁

iSazonov commented:
@NoMoreFood Please check that your new tests fail on current version and do not fail after your fix. Thanks!

NoMoreFood committed:
Add Additional Explicit Encodings For Get-Content -Tail Tests

- Added 'OEM', 'UTF8BOM', and 'UTF8NoBOM' as explicit encodings for existing Get-Content -Tail tests.

NoMoreFood commented:
@iSazonov It appears there are already functional tests for various encodings within the test suite.  However, I added a few additional explicit encodings that were not in the enumeration and committed that change to this pull.  Given this pull addresses a performance issue, I do not see how a reliable pass/fail test can be written given performance can vary drastically between systems and, to a lesser extent, between executions on the same system given other operating system activity.   The only reasonable way to provide a test for something like this would be to add debug/tracing code to the code base to verify the desired code branch is hit/not hit; I do not see any precedent for this type of testing in other tests.

In absence of that, I have provided a before and after demonstration that clearly shows the performance change on my system. 

Test Code:
```
$Encodings = @('String','OEM','Unicode','BigEndianUnicode',
    'UTF8','UTF8BOM','UTF8NoBOM','UTF7','UTF32','Ascii')
$TempFile = (New-TemporaryFile).FullName
$Results = @()
ForEach ($Encoding in $Encodings)
{
    (1..2e6) | Set-Content -Encoding $Encoding -LiteralPath $TempFile -Force
    $Time = Measure-Command { $Capture = Get-Content -Tail 1 -LiteralPath $TempFile }
    $Results += New-Object PSObject -Property @{'Encoding'=$Encoding;'Time'=$Time.TotalMilliSeconds}
    Remove-Item -LiteralPath $TempFile -Force
}
$Results | Format-Table -AutoSize
```

Before Changes:
```
Encoding             Time
--------             ----
String            12.3904
OEM              585.5496
Unicode            1.2809
BigEndianUnicode   1.1258
UTF8             555.5736
UTF8BOM            0.9639
UTF8NoBOM        647.7248
UTF7             560.4215
UTF32              1.1847
Ascii            562.1341
```

After Changes:
```
Encoding           Time
--------           ----
String           2.8139
OEM              0.8078
Unicode          0.8523
BigEndianUnicode 0.9938
UTF8             0.7483
UTF8BOM          0.7559
UTF8NoBOM        0.7721
UTF7             0.8044
UTF32            1.1415
Ascii            0.9463
```

Notice the difference the timings for the code that will be detected as UTF-8 (given the current character set used in the demonstration).  The "after" results are similar to what you would see with Windows PowerShell.

iSazonov commented:
@NoMoreFood I guess the updated test does not fail on current version. Can you check? If so I suggest replace the test text (really it is one byte ASCII)
```c#
        $content = @"
one
two
foo
bar
baz
"@
```
with a text having _multi byte Unicode_, ex.:
```c#
        $content = @"
один
два
фуу
бар
база
"@
```

NoMoreFood commented:
@NoMoreFood Other parts of that test case will need to change as well.  It'll beef it up later today and submit an update.  On the surface after a quick mod, it does not appear to behave any differently before or after these changes (except for the fact it's faster).  More to come...

NoMoreFood commented:
@iSazonov Other parts of that test case will need to change as well to support multi-byte Unicode validation. It'll beef it up later today and submit an update. On the surface after a quick mod, it does not appear to behave any differently before or after these changes (except for the fact it's faster). More to come...

NoMoreFood committed:
Add Multi-Byte Unicode Tail Character Tests

- Modified -Tail encoding test to use three different test sets: utf-8, utf-16, utf-32.  The test verifies that the content resulting from -Tail is equal to the same string returned from a regular Get-Content using both an explicit and implicit encoding.

NoMoreFood commented:
@iSazonov Enhanced Get-Content / Get-Content -Tail tests have been added.

daxian-dbw commented:
@NoMoreFood Is there an issue filed for this fix? If not, can you please open an issue that describes the problem with repro steps?

NoMoreFood commented:
@daxian-dbw Yes, it's in the PR description: https://github.com/PowerShell/PowerShell/issues/11830

daxian-dbw commented:
Oh, I see. Maybe that part of the template is not clear enough :) That's for the doc issue if a documentation change is needed
![image](https://user-images.githubusercontent.com/127450/76639221-fc962c80-650a-11ea-8603-04ff0e6ea3d4.png)

-----------

The issue which the PR is addressing should be put in the PR description, like `Fix #xxxx`. I added it to your PR description.

daxian-dbw commented on src/System.Management.Automation/namespaces/FileSystemContentStream.cs:
```diff
@@ -1460,10 +1460,9 @@ private int RefillByteBuff()
+                _currentEncoding is ASCIIEncoding ||
                 IsSingleByteCharacterSet())
             {
                 // Unicode -- two bytes per character
```
Comment needs to be updated -- remove the `// BigEndianUnicode` part.

NoMoreFood replied:
Done

daxian-dbw reviewed (approved):
LGTM except for one comment.
Thanks for your contribution!

NoMoreFood committed:
Remove BigEndianUnicode Reference In Comment

iSazonov commented:
@NoMoreFood Thanks for your contribution!
